File: src/129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
/*
最基本的方法－深度遍历，保存每一层的结果，碰到叶子存储路径
time defeat: 82%
space defeat: 9%
time consuming: 30 mins
*/
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        //极端情况
        if(root == NULL){
            return 0;
        }
        else if(root->left == NULL && root->right == NULL){
            return root->val;
        }
        else{
            vector<int> path;   //用来存储每条路径的数组
            vector<int> temp;   //用来存储一条路径的数组
            recDfs(root, temp, path);
            /*
            for(auto i : path){
                cout<<i<<endl;
            }*/
            return accumulate(path.begin(), path.end(), 0);
        }
    }
    
    void recDfs(const TreeNode * _node, vector<int> & _nowPath, vector<int> & _allPath){
        if(_node == NULL){
            return;
        }
        else if(_node->left == NULL && _node->right == NULL){
            //碰到叶子节点
            _nowPath.emplace_back(_node->val);
            //组合出当前数字
            string temp = "";
            for(int i: _nowPath){
                temp += to_string(i);
            }
            //压入结果
            _allPath.emplace_back(atoi(temp.c_str()));
            //回退当前结果
            _nowPath.pop_back();
            return;
        }
        else{
            _nowPath.emplace_back(_node->val);
            //深度遍历
            recDfs(_node->left, _nowPath, _allPath);
            recDfs(_node->right, _nowPath, _allPath);
            //释放当前节点
            _nowPath.pop_back();
            return;
        }
    }
};

File: src/129.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        #干翻字节！
        #下星期开始干翻阿里
        #和113一样，递归+回溯
        #不用边界处理
        #很熟练，十二分钟，一次提交，时间击败-96.93%，空间击败-20.83%
        result = [0]  #结果，题目保证正整数，使用列表进行参数传值
        aPath = 0   #一条路径，用整数的化这样就不用回退
        def getPathSum(_node, _path, _result):
            if _node == None:
                return
            elif _node.left == None and _node.right == None:
                #叶节点
                _path = (_path * 10) + _node.val
                #加到结果了
                _result[0] += _path
                return
            else:
                #非叶节点
                _path = (_path * 10) + _node.val
                getPathSum(_node.left, _path, _result)
                getPathSum(_node.right, _path, _result)
                return

        getPathSum(root, aPath, result)
        return result[0]
